extends KinematicBody2D #æ‰©å¼ /å»¶ä¼¸
#const ä¸ºè®¾ç½®å¸¸/å˜é‡
const PlayerHurtSound = preload("res://Player/PlayerHurtSound.tscn")

export var ACCELERATION = 500 #åŠ é€Ÿåº¦ #ä½¿ç”¨exportå¯ä»¥è‡ªç”±è®¾å®šæ•°å€¼ï¼Œå˜é‡å‡ºç°åœ¨å³è¾¹inspectorçª—å£å¯ä»¥è‡ªç”±è°ƒæ•´
export var MAX_SPEED = 80 #æœ€å¤§é€Ÿåº¦
export var ROLL_SPEED = 125 #ç¿»æ»šé€Ÿåº¦
export var FRICTION = 500 #æ‘©æ“¦åŠ›

enum {                          
	MOVE,
	ROLL,
	ATTACK
}#enumä¸ºåˆ—ä¸¾í ¼å‹ç›¸å½“ä¸ä¸€ä¸ªå¸¸æ•°é›†è¿™é‡Œç”¨æ¥è¡¨ç¤ºå¤šä¸ªä¸åŒçš„çŠ¶æ€ï¼Œä»¥ä¾¿è®¾ç½®çŠ¶æ€æœº
var state = MOVE
var velocity = Vector2.ZERO #varä¸ºå®£è¨€å˜é‡ è¿™é‡ŒæŒ‡é€Ÿåº¦ç­‰äº2dç¯å¢ƒä¸‹çš„æ•°å­¦å‘é‡å³å˜é‡é€Ÿåº¦ä¸º0
var roll_vector = Vector2.DOWN #å®£è¨€ç¿»æ»šçš„æ–¹å‘ï¼Œéšç€ç©å®¶æ–¹å‘å˜åŒ–è€Œå˜åŒ–ï¼Œè¿™é‡Œçš„downåº”è¯¥æ˜¯æ¸¸æˆå¼€å§‹æ—¶å€™çš„æ–¹å‘
var stats = PlayerStats

onready var animationPlayer = $AnimationPlayer #ä½¿ç”¨onrready var è¿™ä¸ªå˜é‡å³åªä¼šåœ¨è¿™ä¸ªèŠ‚ç‚¹åˆå§‹åŒ–å®Œæˆæ—¶è¢«åˆ›å»º
							   #ä½¿ç”¨$å¯ä»¥è·å–èŠ‚ç‚¹çš„å˜é‡
onready var animationTree = $AnimationTree
onready var animationState = animationTree.get("parameters/playback") #getä¸ºå–è¿™ä¸ªå˜é‡å–å¾—ä¹‹åæ‰å¯ä»¥è®¾ç½®åŠ¨ç”»
onready var swordHitbox = $HitboxPivot/SwordHitbox #è·å–èŠ‚ç‚¹å˜é‡ï¼Œéœ€è¦ç”¨äºè®¾å®šknockback_vectorç­‰äºæˆ‘ä»¬çš„ç§»åŠ¨æ–¹å‘
onready var hurtbox = $Hurtbox
onready var blinkAnimationPlayer = $BlinkAnimationPlayer

func _ready():      #è®¾ç½®æ­¤é¡¹ååŠ¨ç”»æ ‘å°±åªåœ¨æ¸¸æˆå¼€å§‹æ—¶æ‰ä¼šæ¿€æ´»
	randomize() #è®©æ¸¸æˆæ¯æ¬¡è¿è¡Œéƒ½æœ‰ä¸ä¸€æ ·çš„ç»“æœ
	stats.connect("no_health", self, "queue_free") #é“¾æ¥statsçš„ä¿¡å·
	animationTree.active = true #å³activeçš„å¼€å…³ï¼Œè¿™é‡Œè®¾ç½®ä¸ºtrueåé¢æ¿å¤„å°±éœ€è¦off
	swordHitbox.knockback_vector = roll_vector #å‡»é€€å‘é‡ç­‰äºç¿»æ»šå‘é‡

func _physics_process(delta): #funcä¸ºå‡½æ•°/é–¢æ•°
			#_process çš„æ‰§è¡Œæ—¶é—´æ˜¯åœ¨æ¯æ¬¡å¸§ç»˜åˆ¶å®Œæˆåï¼Œå¯ä»¥ä½¿ç”¨ set_process() æ¥æ‰“å¼€æˆ–è€…å…³é—­ï¼Œ
			#åªè¦å†™äº†è¿™ä¸ªå‡½æ•°ï¼Œé»˜è®¤æ˜¯æ‰“å¼€æ‰§è¡Œçš„ã€‚
			#è€Œ _physics_process åˆ™æ˜¯åœ¨ç‰©ç†å¼•æ“è®¡ç®—ä¹‹å‰æ‰§è¡Œï¼Œè€Œç‰©ç†å¼•æ“çš„è®¡ç®—é—´éš”æ˜¯å›ºå®šçš„ï¼Œæ›´æ–°é¢‘ç‡å¯ä»¥è‡ªè¡Œè®¾å®šã€‚
			#deltaæ˜¯æ¯å¸§ä¹‹é—´çš„ç§’æ•°æ¯”å¦‚20å¸§å°±ä¸º1/20ç§’=0.05deltaï¼Œä¹˜ä»¥deltaå¯ä»¥ä½œä¸ºä¸€ç§åœ¨ä¸åŒå¸§æ•°ä¸‹ä¼šå‡ºç°ä¸åŒæ•ˆæœçš„è§£å†³æ–¹æ¡ˆ
			#å…¬å¼ä¸ºå˜é‡ä¹˜ä»¥delitaåœ¨ä¹˜ä»¥å¸§æ•°ï¼Œæ‰€ä»¥æŠŠå˜é‡ä¹˜ä»¥deltaçš„è¯å°±å¯ä»¥è§£å†³ä»¥ä¸Šé—®é¢˜ã€‚
	match state: #matchç›¸å½“äºå…¶ä»–è¯­è¨€é‡Œçš„switch,ä¸»è¦åŒºåˆ«æ•°è¿™äº›caseå¯ä»¥æ˜¯å˜é‡
	#çŠ¶æ€æœºçš„æ ¸å¿ƒå°±å„ä¸ªçŠ¶æ€çš„ä»£ç åˆ†å¼€è¿›è¡Œå¤„ç†ä¸è½¬æ¢æ˜“äºç®¡ç†å¯»æ‰¾bugç­‰ï¼Œè¿™é‡Œä½¿ç”¨matchè¯­å¥ä¼šå†³å®šå“ªä¸€è¡Œä»£ç è¢«è¿è¡Œã€‚
		MOVE:#è¿™é‡Œç›´æ¥å¼•ç”¨åˆ›å»ºå¥½çš„move_stateè®°å¾—deltaä¹Ÿè¦ä¸€èµ·ä½¿ç”¨
			move_state(delta)

		ROLL:
			roll_state(delta)
			
		ATTACK:
			attack_state(delta)
			
func move_state(delta):
	
	var input_vector = Vector2.ZERO  #å®£è¨€è¾“å…¥é€Ÿåº¦ç­‰äº2dç¯å¢ƒä¸‹çš„æ•°å­¦å‘é‡
	input_vector.x =  Input.get_action_strength("ui_right") - Input.get_action_strength("ui_left")
	#è¾“å…¥Xè½´ï¼ˆæ¨ªè½´ï¼‰ç­‰äº
	#get_action_strengthä¸ºè¿”å›ä»‹äº0å’Œ1ä¹‹é—´çš„å€¼ï¼Œä»£è¡¨ç»™å®šåŠ¨ä½œçš„å¼ºåº¦ã€‚
	#ä¾‹å¦‚ï¼Œåœ¨æ¸¸æˆæ‰‹æŸ„ä¸­ï¼Œè½´ï¼ˆæ¨¡æ‹Ÿæ‘‡æ†æˆ–L2ï¼ŒR2è§¦å‘å™¨ï¼‰ç¦»æ­»åŒºçš„è·ç¦»è¶Šè¿œï¼Œè¯¥å€¼å°±è¶Šæ¥è¿‘1ã€‚
	#å¦‚æœå°†åŠ¨ä½œæ˜ å°„åˆ°æ²¡æœ‰è½´ä½œä¸ºæ§ä»¶çš„æ§ä»¶ï¼Œé”®ç›˜ï¼Œè¿”å›çš„å€¼ä¸º0æˆ–1ã€‚
	input_vector.y = Input.get_action_strength("ui_down") - Input.get_action_strength("ui_up")
	input_vector = input_vector . normalized() #å°†æ–œä¸Šæ–œä¸‹çš„é€Ÿåº¦é™åˆ¶åœ¨åœ†çš„åŠå¾„å†…ï¼Œå³ç­‰äºåŠå¾„
	
	if input_vector != Vector2.ZERO: #å¦‚æœçš„æ¡ä»¶
		roll_vector = input_vector #ç¿»æ»šè·ç¦»ç­‰äºè¾“å…¥æ–¹å‘
		swordHitbox.knockback_vector = input_vector #å‡»é€€è·ç¦»ç­‰äºè¾“å…¥æ–¹å‘
		animationTree.set("parameters/Idle/blend_position", input_vector)#ä½¿ç”¨setæ¥è®¾å®šè·¯å¾„åœ¨èŠ‚ç‚¹å¯ä»¥æŸ¥çœ‹ï¼Œæ³¨æ„å¤§å°å†™ã€‚
		animationTree.set("parameters/Run/blend_position", input_vector)
		animationTree.set("parameters/Attack/blend_position", input_vector)
		animationTree.set("parameters/Roll/blend_position", input_vector)
		animationState. travel("Run")
		velocity = velocity.move_toward(input_vector * MAX_SPEED,ACCELERATION * delta)
		## move towardçš„å‚æ•°åŸºæœ¬æ˜¯å›ºå®šçš„, ä½†æ˜¯éšç€velocityçš„å€¼ä¸æ–­å¢åŠ 
		# ç¬¬ä¸€æ¬¡å‚æ•°ä¸º1 x 100å’Œ 0.016 x 100, æ­¥è¿›1.6, é‚£ä¹ˆç¬¬ä¸€æ¬¡å¾—åˆ°çš„å€¼å°±æ˜¯1.667 
		# ç¬¬äºŒæ¬¡å‚æ•°ä¸å˜, ä½†æ˜¯åˆå§‹å˜äº†, ç›¸å½“äºä»1.6667 åˆ° 100 , deltaæ­¥è¿› 1.6667,é‚£ä¹ˆå¾—åˆ°çš„å€¼å°±ä¸º3.333
		# è¿™æ ·å°±èƒ½ä¿è¯äº†è¿™ä¸ªå‘é‡å€¼(é€Ÿåº¦),æ˜¯è§„å¾‹çš„ä¸æ–­å¢å¤§
		# è¿™é‡Œæœ‰ä¸€ä¸ªå°çŸ¥è¯†ç‚¹,å°±æ˜¯move_towardçš„ä¸¤ä¸ªå€¼(fromå’Œto)çš„å·®å°äºdeltaçš„æ—¶å€™,é‚£ä¹ˆç›´æ¥è¿”å›ç›®çš„(to)çš„å€¼
		# æ¯”å¦‚ var tv3 = move_toward(50, 100, 51) # å¾—åˆ°100
		# è¿™æ ·å°±ä½¿å¾—è¿™ä¸ªå‡½æ•°å†ç‰¹å®šçš„æ—¶å€™å¯ä»¥æœ‰æé™çš„å€¼,ä¹Ÿå°±æ˜¯è¿™é‡Œè®¾å®šçš„æœ€å¤§é€Ÿåº¦
	else: #å¦å¤–çš„æ¡ä»¶ï¼Œå¦‚æœä¸æ»¡è¶³ifçš„æ¡ä»¶å°±æ‰§è¡Œæ­¤æ¡ä»¶
		animationState.travel("Idle")
		velocity = velocity.move_toward(Vector2.ZERO, FRICTION * delta )
	
	move() #åº”ç”¨äºæ•´ä¸ªmoveçŠ¶æ€æœºçš„ä½ç§»å±æ€§
	
	if Input.is_action_just_pressed("roll"):#æŒ‰ä¸‹æŒ‰é”®
		state = ROLL
	
	if Input.is_action_just_pressed("attack"): #æŒ‰ä¸‹æŒ‰é”®
		state = ATTACK

func roll_state(_delta): #ç¿»æ»šçŠ¶æ€æœºä¸‹çš„å¤„ç†
	velocity = roll_vector * ROLL_SPEED  #ç¿»æ»šé€Ÿåº¦çš„è®¾å®šä¸Šé¢å·²ç»äº‹å…ˆåˆ›å»ºå¥½å˜é‡
	animationState.travel("Roll")  #å–ç¿»æ»šçŠ¶æ€æœºçš„åŠ¨ç”»
	move() #é€šè¿‡åŠ å…¥moveå±æ€§å®ç°ä½ç§»

func move():
	velocity = move_and_slide(velocity)  #ä¼šåœ¨å‡½æ•°å†…éƒ¨è‡ªåŠ¨å¤„ç†deltaï¼Œé è¿‘å¢™æ—¶ä¿æŒé€Ÿåº¦ä¸å˜

func attack_state(_delta): #æ”»å‡»çŠ¶æ€ä¸‹çš„å¤„ç†
	velocity = Vector2.ZERO #é˜²æ­¢æ”»å‡»ä½ç§»
	animationState.travel("Attack") #å–æ”»å‡»çš„çŠ¶æ€æœºçš„åŠ¨ç”»

func roll_animation_finished(): #åŠ¨ç”»æ’­æ”¾å®Œæˆåçš„çŠ¶æ€
	velocity = Vector2.ZERO   #é˜²æ­¢ç¿»æ»šåå‡ºç°æ»‘åŠ¨
	state = MOVE

func attack_animation_finished(): #åŠ¨ç”»æ’­æ”¾å®Œæˆåçš„çŠ¶æ€
	state = MOVE

func _on_Hurtbox_area_entered(area):
	stats.health -= area.damage
	hurtbox.start_invincibility(0.6)  #æ— æ•Œæ—¶é—´
	hurtbox.create_hit_effect()
	var playerHurtSounds = PlayerHurtSound.instance()
	get_tree().current_scene.add_child(playerHurtSounds)

func _on_Hurtbox_invincibility_started():
	blinkAnimationPlayer.play("Start")


func _on_Hurtbox_invincibility_ended():
	blinkAnimationPlayer.play("Stop")
